lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -2,6 +2,127 @@
 
 module PolicyMachineStorageAdapter
   class ActiveRecord
+    class PolicyElementAssociation
+      def self.all_accessible_objects(associations, root_id:, filters: {})
+        pea_ids = associations.pluck(:id)
+        return [] if pea_ids.empty?
+
+        query = <<-SQL
+          id IN (
+            WITH RECURSIVE ancestor_scope AS (
+            SELECT
+                id AS child_id,
+                id AS parent_id
+            FROM policy_elements
+            WHERE id = ?
+
+            UNION ALL
+
+            SELECT
+                assignments.child_id,
+                assignments.parent_id
+            FROM assignments
+            JOIN ancestor_scope
+                ON assignments.child_id = ancestor_scope.parent_id
+            ),
+
+            leaf_ancestors AS (
+            SELECT DISTINCT
+                ancestor_scope.parent_id AS parent_id
+            FROM ancestor_scope
+            LEFT OUTER JOIN assignments
+                ON ancestor_scope.parent_id = assignments.child_id
+            WHERE assignments.child_id IS NULL
+            ),
+
+            leaf_descendants AS (
+            SELECT
+                parent_id,
+                parent_id AS child_id
+            FROM leaf_ancestors
+
+            UNION ALL
+
+            SELECT
+                assignments.parent_id,
+                assignments.child_id
+            FROM assignments
+            JOIN leaf_descendants
+                ON leaf_descendants.child_id = assignments.parent_id
+            ),
+
+            accessible_leaf_descendants AS (
+            SELECT DISTINCT
+                leaf_descendants.child_id
+            FROM leaf_descendants
+            JOIN policy_element_associations peas
+                ON peas.object_attribute_id = leaf_descendants.child_id
+            WHERE 1=1
```

Comment:
Comment please.

Reply:
fine, fine, i'll remove more of my query quirks

Reply:
If this come with a performance improvement because of some weird psql thing then we should keep it (and add a comment). I just can't remember if that's the case.

Reply:
nope, just a quirk in how i write SQL.